bot/decorators.py

```diff
@@ -92,6 +95,42 @@ async def inner(self: Cog, ctx: Context, *args, **kwargs) -> None:
     return wrap
 
 
+def mutually_exclusive(namespace: t.Hashable, resource_arg: ResourceId) -> t.Callable:
+    """
+    Turn the decorated coroutine function into a mutually exclusive operation on a resource.
+
+    If any other mutually exclusive function currently holds the lock for a resource, do not run the
+    decorated function and return None.
+
+    `namespace` is an identifier used to prevent collisions among resource IDs.
+
+    `resource_arg` is the positional index or name of the parameter of the decorated function whose
+    value will be the resource ID. It may also be a callable which will return the resource ID
+    given the decorated function's args and kwargs.
+    """
+    def decorator(func: t.Callable) -> t.Callable:
+        @wraps(func)
+        async def wrapper(*args, **kwargs) -> t.Any:
+            if callable(resource_arg):
+                # Call to get the ID if a callable was given.
+                id_ = resource_arg(*args, **kwargs)
+            else:
+                # Retrieve the ID from the args via position or name.
+                id_ = _get_arg_value(resource_arg, args, kwargs)
+
+            # Get the lock for the ID. Create a Lock if one doesn't exist yet.
+            locks = __lock_dicts[namespace]
+            lock = locks.setdefault(id_, asyncio.Lock)
+
+            if not lock.locked():
+                # Resource is free; acquire it.
+                async with lock:
```

Comment:
I created a minimal example and tried to actually get a race condition. While I still think it's possible in theory, since there's a chain of `awaits` before the sentinel `_locked` gets set to `True` in the `asyncio.Lock` instance, it does not actually seem to have the chance to create issues. We may still decide to just ditch it, since we don't need waiting logic (it's what we're trying to prevent, I think), but hmm.

Minimal example I used:
```py
from functools import wraps
import asyncio


locks = {}


def mutually_exclusive(func):
    """Minimal example of race condition."""
    @wraps(func)
    async def wrapper(example_id):
        id_ = 1
        # Get the lock for the ID. Create a lock if one doesn't exist yet.
        lock = locks.setdefault(id_, asyncio.Lock())

        if not lock.locked():
            async with lock:
                return await func(example_id)
        else:
            return 0

    return wrapper


@mutually_exclusive
async def example(example_id):
    print(f"Hello from {example_id}")
    await asyncio.sleep(1)
    print(f"Goodbye from {example_id}")
    return 1


async def main():
    tasks = 100_000
    print("Scheduling all coroutines")
    ran = sum(await asyncio.gather(*[example(i) for i in range(tasks)]))
    print(f"Done, ran {ran} tasks, stopped {tasks-ran} tasks")


asyncio.run(main())
```

Reply:
A closer inspection of the source code suggests that the race condition may not occur, since the first point where the `await` chain will be suspended is after setting the sentinal `self._locked` to `True`. That said, I'm not entirely sure if my understanding of how scheduled tasks are suspended is good enough to figure this out. I think it would work like I think it does if it were `yield from` chains, which suspend at the inner `yield` subgenerator.

Reply:
The decorator coroutine can't yield control on synchronous code. Once one decorator starts executing, any other decorator (actually, the entire program) will be completely blocked until the first decorator reaches an await somewhere, like in `id_ = await id_` or when the lock is acquired. If there was an await between the if condition and the `async with`, then that could indeed cause a race condition. But this is not the case here since it immediately tries to enter the context manager.

Reply:
But you're right, a sentinel is more appropriate. I just didn't feel like creating a custom object when the lock can do the job too. Do you think it's worth using your example to be clearer?

Reply:
> The decorator coroutine can't yield control on synchronous code. Once one decorator starts executing, any other decorator (actually, the entire program) will be completely blocked until the first decorator reaches an await somewhere, like in `id_ = await id_` or when the lock is acquired. If there was an await between the if condition and the `async with`, then that could indeed cause a race condition. But this is not the case here since it immediately tries to enter the context manager.

The line that acquires the lock:
```py
async with lock:
```
*can* yield back control to event loop, as it has multiple awaits within it, like `await self.acquire()`. This means that if every await has the potential to yield back control, control may be yielded back before the `self._locked = True` is run.

Now, I think that because each coroutine is primed when it's scheduled (it runs to the first await), we're still safe, but I'm not entirely sure if the `yield from` comparison/analogy holds.

Reply:
To make the implicit `await` in the context manager a bit more explicit, this is the equivalence code, according the docs:

```py
await lock.acquire()  # <- this is where the await is hiding
try:
    return await func(example_id)
finally:
    lock.release()
```

With the context manager, there are two coroutines involved in the context manager: `__aenter__`, which then awaits `Lock.acquire`.

Reply:
That's correct, but this works based on the assumption that the lock will be immediately acquired if it's available, which is correct:

```py
    async def acquire(self):
        if (not self._locked and (self._waiters is None or
                all(w.cancelled() for w in self._waiters))):
            self._locked = True
            return True
```

I get your question now. I'm not sure, but I don't think that merely an `await` can yield control. I think the coroutine has to actually be doing something asynchronous, like using a future.

Reply:
> I get your question now. I'm not sure, but I don't think that merely an await can yield control. I think the coroutine has to actually be doing something asynchronous, like using a future.

That's the main question I had, yes. If my generator-based analogy with `yield from` holds, scheduling a task would mean running it to the inner most `await` that is waiting for a future. If that's the case, our acquisition of the lock would happen synchronously, as we hit upon this `if` before any such suspension of the task happens. I'm just not entirely sure that is how it works.

Reply:
I've gone ahead with your suggestion to use a custom lock class.